keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,343 @@
+
+- `curReplicas = 10`
+- `calculatedReplicas = 20`
+
+The user specifies only one HPA parameter `constraints.scaleUp.pods = 5` and expects that number of replicas to be set to 15 during the next HPA controller loop.
+But the algorithm picks the softest (largest in this context) change instead:
+
+    Constraint1 = 10 * 2 = 20 (as constraints.scaleUp.percent = 100 by default)
+    Constraint2 = 10 + 5 = 15 (as constraints.scaleUp.pods = 5, set by the user)
+    scaleUpLimit = max(20, 15) = 20
+    desiredReplicas = 20
+
+The user might expect that the autoscaler would use the smallest constraint (15), not the largest one (20). This is not intuitive, but it does make sense if considered thoroughly.
+
+The main idea of the HPA is to autoscale because of a load increase to avoid request failures. It should work on both small clusters and large ones. For small clusters, the absolute number constraint works better (ScaleUpPods), for large clusters the percentage works better (ScaleUpPercentage).
+
+Example: If the current cluster size is `1` and calculated cluster size for this particular load is `20`, then we should reach it ASAP.
+
+For default values (ScaleUpPercent = 100, ScaleUpPods = 4) and “pick the softest constraint” concept, we’ll increase 1 -> 20 in 3 steps
+
+    1 -> 5 -> 10 -> 20
+
+The first step will use the “scaleUp.pods” constraint; the next steps will use “scaleUp.percent” one.
+
+In case of more intuitive “pick the hardest limit” concept, we’ll increase the cluster in 6 steps:
+
+    1 -> 2 -> 4 -> 8 -> 12 -> 16 -> 20
+
+Given that each step takes [90 sec in worst case], we’ll respond to the load increase in `(6-3)*90 sec = ~ 5 min`.
+
+That’s too much, we should respond faster than that.
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+#### Stabilization Window
+
+Stabilization window ([PR][], [RFC][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and a new number of replicas is set to the maximum of all recommendations.
+It will be applied after all the “scaleDown parameters” described above.
+We may want to control stabilization window size in addition to configuring the scale velocity.
+However, at the moment, the “stabilization window feature” is considered to be an internal feature and is not exposed via any configuration.
+
+I’d suggest allowing to change the window size or turn it off if needed, but it should be discussed in a separate RFC.
+
+[PR]: https://github.com/kubernetes/kubernetes/pull/68122
+[RFC]: https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM/edit#heading=h.3tdw2jxiu42f
+
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HPAScaleConstraintValue struct {
+    Pods          int32
+    Percent       int32
+    PeriodSeconds int32
```

Comment:
`PeriodSecond` might not be applicable to everything we want to put in a scale constraint.  For example, if we want to add `Delay` to support custom values for `--horizontal-pod-autoscaler-downscale-stabilization` there isn't an obvious place to put it.  Generally limits might be a delay (hysteresis) or a rate limit, which required a period.

I recommend putting the rate limiting parameters in a `Delta` structure so delay parameters can live alongside them.

E.g.
```golang
type HPAScaleConstraintValue struct {
    Delta        HPAScaleDelta
    DelaySeconds int32
}
type HPAScaleDelta struct {
    Pods          int32
    Percent       int32
    PeriodSeconds int32
}
type HorizontalPodAutoscalerSpec struct {
    ScaleTargetRef CrossVersionObjectReference
    MinReplicas    *int32
    MaxReplicas    int32
    Metrics        []MetricSpec
    ScaleUp        HPAScaleConstraintValue
    ScaleDown      HPAScaleConstraintValue
}
```

I think your idea of implementing the [Stabilization Window](https://github.com/kubernetes/enhancements/blob/d6a61afe99edcd327738d63164ad88ed6c7593f2/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md#stabilization-window) as another parameter is a good one.  And I would like to move it forward, so I'm just making sure we have a place to put it.  :)

Reply:
Maybe `Delta` should be `Rate`.

Reply:
Hm, that is an interesting idea!
Adding hysteresis for scaling up/down will make the stabilization window parameter unneeded at all. You may configure the delay=5min for scaling down and the behavior will be the same.
The same idea is applicable to scaling up, though, will be used rarely for obvious reasons :)

What would you say?

Reply:
> Adding hysteresis for scaling up/down will make the stabilization window parameter unneeded at all.

That's the idea.  The commandline flag can still be useful to provide a default value for DelaySeconds within ScaleDown.

The optional parameters in my structures above should be pointers as well.  So if DelaySeconds is nil, the commandline flag is used (only within the ScaleDown structure).  If 0+ it is valid and used instead of the commandline flag.

Reply:
Yeah, great idea, I will rework my propostal.

Reply:
Done, could you check it: https://github.com/kubernetes/enhancements/pull/883/commits/fe36d8a8699beb19bd7fab423bc12c213dd28cb2